GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m

```diff
@@ -475,6 +602,27 @@ - (void)addEventToTrackingCollections:(GDTCOREvent *)event {
   _targetToEventSet[target] = events;
 }
 
+- (nullable NSNumber *)nextBatchID {
+  __block int32_t lastBatchID = -1;
+  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+  [self libraryDataForKey:gBatchIDCounterKey
+               onComplete:^NSData *_Nonnull(NSData *_Nullable data, NSError *_Nullable error) {
+                 if (!error) {
+                   [data getBytes:(void *)&lastBatchID length:sizeof(int32_t)];
+                 }
+                 if (data == nil) {
+                   lastBatchID = 0;
+                 }
+                 dispatch_semaphore_signal(sema);
+                 int32_t incrementedValue = lastBatchID + 1;
+                 return [NSData dataWithBytes:&incrementedValue length:sizeof(int32_t)];
+               }];
+  if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_SEC)) != 0) {
```

Comment:
IMO it would be much safer to have async method semantic and avoid blocking the caller even for 10 sec. All callers use completion to return the result so it is not a huge issue

Reply:
I'll make this change, but it's going to lead to onComplete hell.

Reply:
yeah, onComplete hell is annoying. We use a Google promises lib in FIS, Crashlytics (and couple more as far as I remember) to deal with it. It's pretty solid and small, maybe it's worth considering it as a dependency to GDT as well?

Reply:
Sounds like a good first project for you ;-)

Reply:
Made async.